### PR TITLE
[Merged by Bors] - feat: changeLevel for Dirichlet characters

### PR DESCRIPTION
**Pull request by mo271**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mo271** [comment]
@MichaelStollBayreuth adding you here, because of the helpful suggestions in #7010.

File: Mathlib/Data/ZMod/Units.lean
Comment (mo271) [inline Mathlib/Data/ZMod/Units.lean:1]: I didn't find a good place for this, so I made a new file. We already have `Data.Int.Units` and `Data.Nat.Units`, so why not also `Data.ZMod.Units`.

**Comment by alexjbest** [comment]
Can we add some "obvious" lemmas about change level here, like `change_level_self` for `n | n` is the identity map, and `change_level_trans` saying that if `n | m | r` then changing level twice is the same as changing just once

**Comment by mo271** [comment]
> Can we add some "obvious" lemmas about change level here, like `change_level_self` for `n | n` is the identity map, and `change_level_trans` saying that if `n | m | r` then changing level twice is the same as changing just once

I was going to do that in a follow-up pull request in order to keep them small, but I'm happy to do it here now. Added those.

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 1]
```diff
@@ -49,4 +51,38 @@ lemma periodic {m : ℕ} (hm : n ∣ m) : Function.Periodic χ m := by
...
+    DirichletCharacter R n →* DirichletCharacter R m :=
+  { toFun := fun ψ ↦ MulChar.ofUnitHom (ψ.toUnitHom.comp (ZMod.unitsMap hm)),
+    map_one' := by ext; simp,
+    map_mul' := fun ψ₁ ψ₂ ↦ by ext; simp }
+
+lemma change_level_def {m : ℕ} (hm : n ∣ m) :
+    change_level hm χ = MulChar.ofUnitHom (χ.toUnitHom.comp (ZMod.unitsMap hm)) := rfl
+
+lemma change_level_def' {m : ℕ} (hm : n ∣ m) :
+    (change_level hm χ).toUnitHom = χ.toUnitHom.comp (Units.map (ZMod.castHom hm (ZMod n))) := by
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:66]: Is there a reason why you are not using `ZMod.unitsMap` here?

File: Mathlib/Data/ZMod/Units.lean [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+
+namespace ZMod
+
+variable {n m : ℕ}
+/-- `unitsMap` is a group homomorphism that maps units of `ZMod m` to units of `ZMod n` when `n`
+divides `m`. -/
+def unitsMap (hm : n ∣ m) : (ZMod m)ˣ →* (ZMod n)ˣ := Units.map (ZMod.castHom hm (ZMod n))
+
+lemma unitsMap_def (hm : n ∣ m) : ZMod.unitsMap hm = Units.map (ZMod.castHom hm (ZMod n)) := rfl
```
Comment (MichaelStollBayreuth) [inline Mathlib/Data/ZMod/Units.lean:21]: You can leave out `ZMod` as you are in the namespace here.

File: Mathlib/Data/ZMod/Units.lean [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
```
Comment (MichaelStollBayreuth) [inline Mathlib/Data/ZMod/Units.lean:22]: Maybe add the `trans` lemma here:
```lean
lemma unitsMap_trans {m d : ℕ} (hm : n ∣ m) (hd : m ∣ d) :
  (unitsMap hm).comp (unitsMap hd) = unitsMap (dvd_trans hm hd)
```
and use that for the proof of `changeLevel_trans` later?

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 4]
```diff
@@ -49,4 +51,38 @@ lemma periodic {m : ℕ} (hm : n ∣ m) : Function.Periodic χ m := by
   rw [← ZMod.nat_cast_zmod_eq_zero_iff_dvd] at hm
   simp only [hm, add_zero]
 
+/-- A function that modifies the level of a Dirichlet character to some multiple
+  of its original level. -/
+noncomputable def change_level {R : Type} [CommMonoidWithZero R] {n m : ℕ} (hm : n ∣ m) :
```
[the remark this answers is not in the document]
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: yes, I suppose. Done

**Comment by alexjbest** [comment]
> > Can we add some "obvious" lemmas about change level here, like `change_level_self` for `n | n` is the identity map, and `change_level_trans` saying that if `n | m | r` then changing level twice is the same as changing just once
> 
> I was going to do that in a follow-up pull request in order to keep them small, but I'm happy to do it here now. Added those.

I appreciate that! I think there is some sort of happy medium around 100 lines for a new Def + the basic lemmas to be sure that the Def works as intended.

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 1 above)
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: nope, just forgot to replace this instance of it...

File: Mathlib/Data/ZMod/Units.lean (hunk 2 above)
Comment (mo271) [reply Mathlib/Data/ZMod/Units.lean]: done

File: Mathlib/Data/ZMod/Units.lean (hunk 3 above)
Comment (mo271) [reply Mathlib/Data/ZMod/Units.lean:28]: Good idea! I did that (castHom_comp` was already there...).

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 5]
```diff
@@ -49,4 +51,36 @@ lemma periodic {m : ℕ} (hm : n ∣ m) : Function.Periodic χ m := by
...
+    map_mul' := fun ψ₁ ψ₂ ↦ by ext; simp }
+
+lemma changeLevel_def {m : ℕ} (hm : n ∣ m) :
+    changeLevel hm χ = MulChar.ofUnitHom (χ.toUnitHom.comp (ZMod.unitsMap hm)) := rfl
+
+lemma changeLevel_def' {m : ℕ} (hm : n ∣ m) :
+    (changeLevel hm χ).toUnitHom = χ.toUnitHom.comp (ZMod.unitsMap hm) := by
+  ext
+  rw [changeLevel_def, ZMod.unitsMap_def]
+  simp
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:69]:
```lean
  simp [changeLevel]
```
is enough.

File: Mathlib/Data/ZMod/Basic.lean [hunk 6]
```diff
@@ -1162,9 +1162,14 @@ theorem ringHom_eq_of_ker_eq [CommRing R] (f g : R →+* ZMod n)
 #align zmod.ring_hom_eq_of_ker_eq ZMod.ringHom_eq_of_ker_eq
 
 @[simp]
-lemma castHom_self {n : ℕ} : ZMod.castHom dvd_rfl (ZMod n) = RingHom.id (ZMod n) :=
+lemma castHom_self : ZMod.castHom dvd_rfl (ZMod n) = RingHom.id (ZMod n) :=
   RingHom.ext_zmod (ZMod.castHom dvd_rfl (ZMod n)) (RingHom.id (ZMod n))
```
Comment (MichaelStollBayreuth) [inline Mathlib/Data/ZMod/Basic.lean:1166]: `Subsingleton.elim _ _` works, too.

File: Mathlib/Data/ZMod/Units.lean [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+lemma unitsMap_def (hm : n ∣ m) : unitsMap hm = Units.map (castHom hm (ZMod n)) := rfl
+
+lemma unitsMap_comp {d : ℕ} (hm : n ∣ m) (hd : m ∣ d) :
+  (unitsMap hm).comp (unitsMap hd) = unitsMap (dvd_trans hm hd) := by
+    simp only [unitsMap_def]
+    rw [← Units.map_comp]
+    exact congr_arg Units.map <| congr_arg RingHom.toMonoidHom <| castHom_comp hm hd
+
+end ZMod
```
Comment (MichaelStollBayreuth) [inline Mathlib/Data/ZMod/Units.lean:29]: Perhaps also add
```lean
@[simp]
lemma unitsMap_self (n : ℕ) : unitsMap (dvd_refl n) = MonoidHom.id _ := by
  simp [unitsMap, castHom_self]
```

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 8]
```diff
@@ -49,4 +51,36 @@ lemma periodic {m : ℕ} (hm : n ∣ m) : Function.Periodic χ m := by
...
+
+@[simp]
+lemma changeLevel_self : changeLevel (dvd_refl n) χ = χ := by
+  ext
+  rw [changeLevel_def]
+  simp [ZMod.unitsMap]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:75]: `simp [changeLevel, ZMod.castHom_self]`

File: Mathlib/Data/ZMod/Basic.lean (hunk 6 above)
Comment (mo271) [reply Mathlib/Data/ZMod/Basic.lean]: done

File: Mathlib/Data/ZMod/Units.lean
Comment (mo271) [reply Mathlib/Data/ZMod/Units.lean:1]: done

File: Mathlib/Data/ZMod/Units.lean (hunk 7 above)
Comment (mo271) [reply Mathlib/Data/ZMod/Units.lean:33]: done

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 5 above)
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: done

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 8 above)
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: done

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 9]
```diff
@@ -49,4 +51,36 @@ lemma periodic {m : ℕ} (hm : n ∣ m) : Function.Periodic χ m := by
...
+
+lemma changeLevel_self_toUnitHom : (changeLevel (dvd_refl n) χ).toUnitHom = χ.toUnitHom := by
+  rw [changeLevel_self]
+
+lemma changeLevel_trans {m d : ℕ} (hm : n ∣ m) (hd : m ∣ d) :
+  changeLevel (dvd_trans hm hd) χ = changeLevel hd (changeLevel hm χ) := by
+  simp only [changeLevel_def, toUnitHom_eq, ofUnitHom_eq, Equiv.apply_symm_apply]
+  rw [← ZMod.unitsMap_comp hm hd]
+  rfl
```
[the remark this answers is not in the document]
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: done

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 10]
```diff
@@ -49,4 +51,30 @@ lemma periodic {m : ℕ} (hm : n ∣ m) : Function.Periodic χ m := by
...
+
+@[simp]
+lemma changeLevel_self : changeLevel (dvd_refl n) χ = χ := by
+  simp [changeLevel, ZMod.unitsMap]
+
+lemma changeLevel_self_toUnitHom : (changeLevel (dvd_refl n) χ).toUnitHom = χ.toUnitHom := by
+  rw [changeLevel_self]
+
+lemma changeLevel_trans {m d : ℕ} (hm : n ∣ m) (hd : m ∣ d) :
+  changeLevel (dvd_trans hm hd) χ = changeLevel hd (changeLevel hm χ) := by
```
Comment (alexjbest) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:77]:
```suggestion
    changeLevel (dvd_trans hm hd) χ = changeLevel hd (changeLevel hm χ) := by
```

File: Mathlib/Data/ZMod/Units.lean [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+divides `m`. -/
+def unitsMap (hm : n ∣ m) : (ZMod m)ˣ →* (ZMod n)ˣ := Units.map (castHom hm (ZMod n))
+
+lemma unitsMap_def (hm : n ∣ m) : unitsMap hm = Units.map (castHom hm (ZMod n)) := rfl
+
+lemma unitsMap_comp {d : ℕ} (hm : n ∣ m) (hd : m ∣ d) :
+  (unitsMap hm).comp (unitsMap hd) = unitsMap (dvd_trans hm hd) := by
+    simp only [unitsMap_def]
+    rw [← Units.map_comp]
+    exact congr_arg Units.map <| congr_arg RingHom.toMonoidHom <| castHom_comp hm hd
```
Comment (alexjbest) [inline Mathlib/Data/ZMod/Units.lean:27]:
```suggestion
    (unitsMap hm).comp (unitsMap hd) = unitsMap (dvd_trans hm hd) := by
  simp only [unitsMap_def]
  rw [← Units.map_comp]
  exact congr_arg Units.map <| congr_arg RingHom.toMonoidHom <| castHom_comp hm hd
```

**Comment by MichaelStollBayreuth** [comment]
Should the indentation in
```lean
lemma toUnitHom_eq_char' {a : ZMod n} (ha : IsUnit a) :
  χ a = χ.toUnitHom ha.unit := by simp

lemma toUnitHom_eq_iff (ψ : DirichletCharacter R n) :
  toUnitHom χ = toUnitHom ψ ↔ χ = ψ := by simp

lemma eval_modulus_sub (x : ZMod n) :
  χ (n - x) = χ (-x) := by simp
```
also be changed (to indent by 4 for statements on the line(s) after ":")?

**Comment by mo271** [comment]
> also be changed (to indent by 4 for statements on the line(s) after ":")?

I just put them on one line.

**Comment by MichaelStollBayreuth** [comment]
Looks good to me now!

**Review by alexjbest — COMMENTED** [review]
maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**